mlflow/utils/search_utils.py

```diff
@@ -303,7 +229,93 @@ def does_run_match_clause(cls, run, sed):
         else:
             return False
 
-    def filter(self, run):
-        if not self.parsed:
-            self.parsed = self._parse()
-        return all([self.does_run_match_clause(run, s) for s in self.parsed])
+    @classmethod
+    def _parse_search_filter(cls, filter_string):
+        if not filter_string:
+            return []
+        try:
+            parsed = sqlparse.parse(filter_string)
+        except Exception:
+            raise MlflowException("Error on parsing filter '%s'" % filter_string,
+                                  error_code=INVALID_PARAMETER_VALUE)
+        if len(parsed) == 0 or not isinstance(parsed[0], Statement):
+            raise MlflowException("Invalid filter '%s'. Could not be parsed." %
+                                  filter_string, error_code=INVALID_PARAMETER_VALUE)
+        elif len(parsed) > 1:
+            raise MlflowException("Search filter contained multiple expression '%s'. "
+                                  "Provide AND-ed expression list." % filter_string,
+                                  error_code=INVALID_PARAMETER_VALUE)
+        return SearchUtils._process_statement(parsed[0])
+
+    @classmethod
+    def filter(cls, runs, filter_string):
+        """Filters a set of runs based on a search filter string."""
+        if not filter_string:
+            return runs
+        parsed = cls._parse_search_filter(filter_string)
+
+        def run_matches(run):
+            return all([cls._does_run_match_clause(run, s) for s in parsed])
+        return [run for run in runs if run_matches(run)]
+
+    @classmethod
+    def _parse_order_by(cls, order_by):
+        try:
+            parsed = sqlparse.parse(order_by)
+        except Exception:
+            raise MlflowException("Error on parsing order_by clause '%s'" % order_by,
+                                  error_code=INVALID_PARAMETER_VALUE)
+        if len(parsed) != 1 or not isinstance(parsed[0], Statement):
+            raise MlflowException("Invalid order_by clause '%s'. Could not be parsed." %
+                                  order_by, error_code=INVALID_PARAMETER_VALUE)
+
+        statement = parsed[0]
+        if len(statement.tokens) != 1 or not isinstance(statement[0], Identifier):
+            raise MlflowException("Invalid order_by clause '%s'. Could not be parsed." %
+                                  order_by, error_code=INVALID_PARAMETER_VALUE)
+
+        token_value = statement.tokens[0].value
+        is_ascending = True
+        if token_value.lower().endswith(" desc"):
```

Comment:
do we also need to worry about multiple spaces, tabs ... etc? Or is that normalized by the parser?
I would add some tests for that

Reply:
Tests added. Previously, support for two spaces did not work.

Reply:
curious ... how does the multi space thing work?

Reply:
We trim whitespace after removing the suffix